HP-385 fix bugs and code smells reported by Sonar

NikoHelle opened the pull request:
Original list of bug, security hotspots and code smells: https://sonarcloud.io/dashboard?id=City-of-Helsinki_open-city-profile-ui

Done outside of this PR:
- excluded all generated files from checks
- deactivated rule for  "Shorthand object properties should be grouped at the beginning or end of an object declaration"

NikoHelle committed:
Added linting rule 'curly' and fixed raised errors
Sonar rule: https://sonarcloud.io/organizations/city-of-helsinki/rules?open=typescript%3AS121&rule_key=typescript%3AS121

NikoHelle committed:
Added linting rule 'consistent-return' and fixed raised errors
Sonar rule: https://sonarcloud.io/organizations/city-of-helsinki/rules?open=typescript:S3801&rule_key=typescript:S3801

NikoHelle committed:
Added linting rule 'no-shadow' and fixed raised errors
Sonar rule: https://sonarcloud.io/organizations/city-of-helsinki/rules?open=typescript%3AS1117&rule_key=typescript%3AS1117.

NikoHelle committed:
fix Sonar error The 'any' type should not be used

NikoHelle committed:
Installed eslint-plugin-sonarjs and added linting rule 'no-duplicate-string'
Fixed errors raised by new rule. Rule documentation https://sonarcloud.io/organizations/city-of-helsinki/rules?open=typescript%3AS1192&rule_key=typescript%3AS1192

NikoHelle committed:
Added linting rule 'arrow-body-style' and fixed raised errors
Sonar rule: https://sonarcloud.io/organizations/city-of-helsinki/rules?open=typescript%3AS1192&rule_key=typescript%3AS1192.
Also added rule 'arrow-parens' forcing arrow function format to be consistent and fixed raised errors.
Also fixed type errors with HDSOptionType which were shown only on compile

NikoHelle committed:
Added linting rule 'object-shorthand' and fixed raised errors
Sonar rule: https://sonarcloud.io/organizations/city-of-helsinki/rules?open=typescript%3AS3498&rule_key=typescript%3AS3498.

NikoHelle committed:
Fix rule errors where both '?' and 'undefined' were used
Sonar rule: https://sonarcloud.io/organizations/city-of-helsinki/rules?open=typescript%3AS4782&rule_key=typescript%3AS4782

NikoHelle committed:
Added linting rule '@typescript-eslint/no-unnecessary-type-assertion' and fixed raised errors
Sonar rule: https://sonarcloud.io/organizations/city-of-helsinki/rules?open=typescript%3AS4325&rule_key=typescript%3AS4325. parserOptions in .eslintrc was a requirement for the rule to work.

NikoHelle committed:
Fixed errors for rule 'Default export names and file names should match'
Sonar rule: https://sonarcloud.io/organizations/city-of-helsinki/rules?open=typescript%3AS3317&rule_key=typescript%3AS3317. Also fixed type error discovered when starting app

NikoHelle committed:
Added missing dependency used just for typing

NikoHelle committed:
Fix missing lang, title and integration hash.

Two bugs and a security hotspot reported by Sonar

NikoHelle committed:
Fixed Sonar code smell 'Optional boolean parameters should have default value'

NikoHelle committed:
Moved sonarjs eslint package to dependencies to fix build error in CI

NikoHelle committed:
Fix codesmell 'Redundant casts and non-null assertions should be avoided'

akikoskinen commented on src/profile/helpers/__tests__/updateMutationVariables.test.ts:
```diff
@@ -12,6 +12,9 @@ import { FormValues } from '../../components/editProfileForm/EditProfileForm';
 import { myProfile } from '../../../common/test/myProfileQueryData';
 
+const constantTestEmail = 'kolmas@testi.fi';
+const constantTestAddress = 'Testikatu 66';
```
In my opinion this change (the whole commit bbcca29d3c) doesn't improve the code. When looking at for example this test code change, I think the previous version is better. The [Sonar rule](https://sonarcloud.io/organizations/city-of-helsinki/rules?open=typescript%3AS1192&rule_key=typescript%3AS1192) is very simplistic and doesn't consider the context at all. As such I would even consider disabling the whole rule.

NikoHelle replied:
Fixed properly by using the values from variables where values are initially set.

akikoskinen commented on src/profile/components/editProfileForm/EditProfileForm.tsx:
```diff
@@ -185,7 +185,7 @@ function EditProfileForm(props: Props): React.ReactElement {
       validationSchema={schema}
     >
-      {(formikProps: FormikProps<FormValues>) => (
+      {(editProfileFormikProps: FormikProps<FormValues>) => (
```
This commit (8fb9f7a5ca) doesn't look to me like the right solution. The tool is helpful, it's indicating a code smell -- it says it's code duplication, but it could also be too large components, wrong architecture or a number of other things. Instead of silencing the tool by making worse code, find a solution that **improves** the code. That's anyway the idea in using these tools in the first place.

NikoHelle replied:
Yes, agreed. I removed the "fix" and the code smell can remain. These components that have duplicate code will be refactored anyway when user data from strong authentication is added. Better to fix this then.

akikoskinen commented on src/auth/authService.ts:
```diff
@@ -85,12 +85,13 @@ export class AuthService {
       }
-      return Promise.reject();
+    } finally {
+      Promise.resolve();
```
If error occurs, previously a rejected Promise was returned, now a resolved Promise. That changes the behavior of this function. Is that ok?

NikoHelle replied:
login() causes a full page redirect and none of the callers listen to the result. But returned value should of course be logical. I made some changes to catch a possible error and return rejected / resolved promise in a simpler way.

akikoskinen commented on src/subscriptions/components/subsciptions/Subscriptions.tsx:
```diff
@@ -132,7 +132,7 @@ function Subscriptions(): React.ReactElement {
     index: number,
     code?: string,
-    value?: boolean
+    value = false
```
Can this still also have the explicit type? Something like this perhaps:

    value: boolean = false

NikoHelle replied:
No, it raises an error. Setting a default value also types it. "value" can only be true / false after that.

Error: "Type boolean trivially inferred from a boolean literal, remove type annotation."

akikoskinen replied:
Got it.

akikoskinen reviewed (changes requested):
These one-liner commit messages are really ugly, especially with complex URLs within them. Use line breaks in commit messages.

![Screenshot 2021-01-15 at 14 33 56](https://user-images.githubusercontent.com/758558/104727892-c59e3200-573e-11eb-9c10-213c78ca5f57.png)

akikoskinen commented on src/profile/helpers/__tests__/updateMutationVariables.test.ts:
```diff
@@ -134,7 +134,7 @@ test('add arrays are formed correctly', () => {
+      address: addresses[1].address,
       postalCode: '00000',
       city: 'Helsinki',
       countryCode: 'FI',
```
There's still a lot of repetition even if the tool doesn't complain anymore. Would it be possible to use the `addresses[1]` as a whole (perhaps selecting only wanted fields out of it or something)?

NikoHelle replied:
The original test seems to be done quick & dirty. There's so much repetition. I've now commited massive changes to the tests. They are not squashed yet, because if it is not satisfactory yet, I don't want to squash multiple times.

akikoskinen replied:
Dosn't look that massive to me. But cleaner 😃

NikoHelle committed:
Changed tests to use same, one time configured data

Tests had a lot of data defined multiple times. Matching data is now cloned when needed and set only once. Data objects are cloned because passing original objects can mess tests or allow them pass falsely.

akikoskinen commented on src/profile/helpers/__tests__/updateMutationVariables.test.ts:
```diff
@@ -97,33 +112,12 @@ const formValues: FormValues = {
-    addressType: AddressType.OTHER,
-    __typename: 'AddressNode',
-  },
+  primaryEmail: cloneObject(emails[0]),
```
Could this clone the `primaryEmail`?

NikoHelle replied:
Tests might break, if data not in the arrays (emails, addresses, phones) is used here where test data is created.  "primaryEmail/Phone/Address" are currently used in the array, but that could change or their data could be used with modifications. 

Bottom line: If arrays change, the changes should not break the tests.

akikoskinen commented on src/profile/helpers/__tests__/updateMutationVariables.test.ts:
```diff
@@ -97,33 +112,12 @@ const formValues: FormValues = {
+  primaryEmail: cloneObject(emails[0]),
+  emails,
+  phones,
+  primaryPhone: cloneObject(phones[0]),
```
Could this clone the `primaryPhone`?

NikoHelle replied:
Same comment as above

akikoskinen commented on src/profile/helpers/__tests__/updateMutationVariables.test.ts:
```diff
@@ -97,33 +112,12 @@ const formValues: FormValues = {
+  phones,
+  primaryPhone: cloneObject(phones[0]),
+  addresses,
+  primaryAddress: cloneObject(addresses[0]),
```
Could this clone the `primaryAddress`?

NikoHelle replied:
Same comment as above

akikoskinen commented on src/profile/helpers/__tests__/updateMutationVariables.test.ts:
```diff
@@ -175,30 +162,13 @@ test('update arrays are formed correctly', () => {
   );
 
   expect(variables.input.profile.addAddresses).toEqual([
```
Not related to this PR in any way, but should this be checking a field named `updateAddresses`?

NikoHelle replied:
Good catch. That made no sense and is now fixed.

akikoskinen reviewed (approved):
Few comments still, but looks good already like this. I suppose solving the conflict with the current `develop` branch doesn't change the contents of this PR in any way.

NikoHelle committed:
Rename test data to improve clarity and create test data in cleaner way
Test data is simplified by cloning existing Profile data instead of declaring data multiple times. Naming data instead of using arrays and indexes makes tests easier to understand and follow.

akikoskinen reviewed (approved):
Pretty nice.
